src/Usb.Net/Windows/WindowsUsbInterface.cs

```diff
@@ -71,29 +71,59 @@ public void Dispose()
             GC.SuppressFinalize(this);
         }
 
-        public Task<ReadResult> SendControlTransferAsync(SetupPacket setupPacket, byte[] inputBuffer, CancellationToken cancellationToken = default)
+        public Task<ControlTransferResult> SendControlTransferAsync(SetupPacket setupPacket, byte[] buffer, CancellationToken cancellationToken = default)
         {
             return Task.Run(() =>
             {
-                uint bytesWritten = 0;
+                var transferBuffer = Array.Empty<byte>();
 
-                if (setupPacket.Length > 0 && inputBuffer == null)
+                uint bytesTransferred = 0;
+
+                if (setupPacket.Length > 0)
                 {
-                    inputBuffer = new byte[setupPacket.Length];
+                    // the request has data exchange
+                    // which direction?
+                    if (setupPacket.RequestType.Direction == RequestDirection.In)
+                    {
+                        // data is to be received
+                        // create buffer of the appropriate size
+                        transferBuffer = new byte[setupPacket.Length];
+                    }
+                    else
+                    {
+                        // data is to be sent
+                        // TODO
+                        transferBuffer = new byte[SetupPacket.SetupPacketSize + setupPacket.Length];
+                        // check buffer == null
+                    }
                 }
 
-                //Make a copy so we don't mess with the array passed in
-                var bufferCopy = new byte[inputBuffer.Length];
-                Array.Copy(inputBuffer, bufferCopy, inputBuffer.Length);
+                ////Make a copy so we don't mess with the array passed in
```

Comment:
@josesimoes it's not a good idea to mutate the array that the user passes in.

The windows API shouldn't do that.